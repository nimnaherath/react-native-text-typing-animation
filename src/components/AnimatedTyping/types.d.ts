import { type TextStyle } from 'react-native';

export type Props = {
  text: string;
  style?: TextStyle;
  curserColor?: string;
  cursorBlink?: number;
  duration?: number;
  started?: boolean;
  onCompleted?: () => void;
  textComponent?: React.ComponentType<TextProps>;
};
