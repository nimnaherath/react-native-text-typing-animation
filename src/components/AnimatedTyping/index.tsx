import React, { forwardRef, useImperativeHandle, useState } from 'react';
import { View } from 'react-native';
import { type TextStyle } from 'react-native';

import type { Props } from './types';
import CurserComponent from '../CurserComponent';
import TextComponent from '../TextComponent';

export interface AnimatedTextRef {
  start: () => void;
  pause: () => void;
  isCompleted: boolean;
  clear: () => void;
}

const AnimatedTyping = forwardRef<AnimatedTextRef, Props>(
  (
    {
      text,
      style,
      curserColor = 'black',
      cursorBlink = 100,
      duration = 200,
      started = false,
      onCompleted,
      textComponent: CustomText = TextComponent,
    }: Props,
    ref
  ) => {
    const [animationStarted, setAnimationStarted] = useState(started);
    const [textIndex, setTextIndex] = useState(0);
    const [displayedText, setDisplayedText] = useState('');
    const [curserBlink, setCurserBlink] = useState(true);

    useImperativeHandle(ref, () => ({
      start: () => setAnimationStarted(true),
      pause: () => setAnimationStarted(false),
      isCompleted: text.length <= textIndex,
      clear: () => {
        setTextIndex(0);
        setDisplayedText('');
      },
    }));

    if (animationStarted) {
      if (textIndex <= text.length) {
        setTimeout(() => {
          setDisplayedText(text.substring(0, textIndex));
          setTextIndex(textIndex + 1);
        }, duration);
        setTimeout(() => {
          setCurserBlink(!curserBlink);
        }, cursorBlink);
      } else {
        setAnimationStarted(false);
        if (onCompleted) {
          onCompleted();
        }
      }
    }

    const textStyle: TextStyle = {
      ...style,
    };

    return (
      <View
        style={{
          flexWrap: 'nowrap',
        }}
      >
        <CustomText style={textStyle}>
          {displayedText}
          {animationStarted && (
            <CurserComponent
              curserColor={curserColor}
              curserBlink={curserBlink}
            />
          )}
        </CustomText>
      </View>
    );
  }
);

export default AnimatedTyping;
