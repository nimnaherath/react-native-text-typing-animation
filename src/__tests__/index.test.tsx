import React from 'react';
import renderer from 'react-test-renderer';

import AnimatedTyping from '../index';

jest.useFakeTimers();

describe('animated typing component', () => {
  test('component render correctly', () => {
    const textAnimation = renderer
      .create(<AnimatedTyping text="Test" />)
      .toJSON();

    expect(textAnimation).toMatchSnapshot();
  });

  test('it can change style', () => {
    const textAnimation = renderer
      .create(
        <AnimatedTyping
          text="Test"
          style={{ color: 'black', fontSize: 20, fontWeight: '700' }}
        />
      )
      .toJSON();

    expect(textAnimation).toMatchSnapshot();
  });
});
