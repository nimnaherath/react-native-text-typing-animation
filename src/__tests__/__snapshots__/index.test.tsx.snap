// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`animated typing component component render correctly 1`] = `
<View
  style={
    Object {
      "flexWrap": "nowrap",
    }
  }
>
  <Text />
</View>
`;

exports[`animated typing component it can change style 1`] = `
<View
  style={
    Object {
      "flexWrap": "nowrap",
    }
  }
>
  <Text />
</View>
`;
